lib/authz/project_rules_update_workflow.go

```diff
@@ -0,0 +1,265 @@
+package authz
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	iam_v2 "github.com/chef/automate/api/interservice/authz/v2"
+	"github.com/chef/automate/lib/cereal"
+)
+
+const (
+	cancelUpdateProjectTagsTaskName = "CancelUpdateProjectTags"
+	startProjectTagUpdaterTaskName  = "StartProjectTagUpdater"
+	projectTagUpdaterStatusTaskName = "ProjectTagUpdaterStatus"
+)
+
+type DomainProjectUpdateWorkflowExecutor struct {
+	PollInterval time.Duration
+
+	cancelUpdateProjectTagsTaskName string
+	startProjectTagUpdaterTaskName  string
+	projectTagUpdaterStatusTaskName string
+}
+
+type DomainProjectUpdateWorkflowParameters struct {
+	ProjectUpdateID string
+}
+
+type DomainProjectUpdateWorkflowPayload struct {
+	ProjectUpdateID             string
+	ESJobID                     []string
+	ConsecutiveJobCheckFailures int
+	MergedJobStatus             JobStatus
+}
+
+type statusParameters struct {
+	EsJobIDs []string
+}
+
+func (m *DomainProjectUpdateWorkflowExecutor) OnStart(
+	w cereal.WorkflowInstance, ev cereal.StartEvent) cereal.Decision {
+
+	params := DomainProjectUpdateWorkflowParameters{}
+	if err := w.GetParameters(&params); err != nil {
+		return w.Fail(err)
+	}
+
+	logrus.Debugf("Started DomainProjectUpdateWorkflow for %s",
+		params.ProjectUpdateID)
+	w.EnqueueTask(m.startProjectTagUpdaterTaskName, nil) // nolint: errcheck
+	return w.Continue(DomainProjectUpdateWorkflowPayload{
+		ProjectUpdateID: params.ProjectUpdateID,
+	})
+}
+
+func (m *DomainProjectUpdateWorkflowExecutor) OnTaskComplete(
+	w cereal.WorkflowInstance, ev cereal.TaskCompleteEvent) cereal.Decision {
+
+	payload := DomainProjectUpdateWorkflowPayload{}
+	if err := w.GetPayload(&payload); err != nil {
+		logrus.WithError(err).Errorf("Failed to launch %s", m.cancelUpdateProjectTagsTaskName)
+		return w.Fail(err)
+	}
+
+	switch ev.TaskName {
+	case m.cancelUpdateProjectTagsTaskName:
+		return w.Complete()
+	case m.startProjectTagUpdaterTaskName:
+		esJobIDs := []string{}
+		if err := ev.Result.Get(&esJobIDs); err != nil {
+			logrus.WithError(err).Errorf(
+				"Failed to deserialize %s result", m.startProjectTagUpdaterTaskName)
+			return w.Fail(err)
+		}
+		if len(esJobIDs) <= 0 {
+			err := errors.New("0 ES jobs started")
+			logrus.WithError(err).Errorf(
+				"Failed to deserialize %s result", m.startProjectTagUpdaterTaskName)
+			return w.Fail(err)
+		}
+		w.EnqueueTask(m.projectTagUpdaterStatusTaskName, statusParameters{esJobIDs}, cereal.StartAfter(m.nextCheck())) // nolint: errcheck
```

Comment:
Why is another "start project update" queued after a "start project update" is completed.